compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -1170,6 +1171,52 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             _ => return,
         }
     }
+
+    fn note_fn_method_def_due_to_call(
+        &self,
+        err: &mut Diagnostic,
+        expr: &hir::Expr<'_>,
+        expected: Ty<'_>,
+    ) {
+        let (def_id, ident, callee_str) = if let hir::ExprKind::Call(fun, _) = expr.kind
+            && let hir::ExprKind::Path(hir::QPath::Resolved(_, path)) = fun.kind
+            && let hir::def::Res::Def(def_kind, def_id) = path.res
+            && !matches!(def_kind, hir::def::DefKind::Ctor(..))
+        {
+            (def_id, path.segments[0].ident, "function")
+        } else if let hir::ExprKind::MethodCall(method, ..) = expr.kind
+            && let Some(def_id) = self.typeck_results.borrow().type_dependent_def_id(expr.hir_id)
+            && !matches!(self.tcx.def_kind(def_id), hir::def::DefKind::Ctor(..))
```

Comment:
This last condition is common between both chains.
Should it be put a few lines below, with a set of accepted DefKinds instead of refusing Ctor?
I mean some `let DefKind::Fn | DefKind::AssocFn = tcx.def_kind(def_id) else { return };`

Reply:
Added a separate `if` to match on `DefKind::AssocFn` and `DefKind::Fn`, thanks.